tests/unit/charms/tls_certificates_interface/v4/test_tls_certificates_v4.py

```diff
@@ -69,7 +68,7 @@ def test_given_subject_and_private_key_when_generate_csr_then_csr_is_generated_w
 
     csr = generate_csr(private_key=private_key, common_name=subject)
 
-    csr_object = x509.load_pem_x509_csr(data=csr)
+    csr_object = x509.load_pem_x509_csr(data=csr.encode())
```

Comment:
With my last suggestion, I was kinda hoping we could avoid this type of thing.

We definitely can here, in the tests: we create the `x509.CertificateSigningRequest`, convert it to bytes, then to a string, then to bytes, then back to a `x509.CertificateSigningRequest`. We can solve this pretty easily in the test helper functions by just returning the `x509.CertificateSigningRequest` and removing line 71.

But, I feel like we do this in the actual charm code too, but it's obfuscated by a bunch of helper functions ([does this CSR match? let's convert everything back!](https://github.com/canonical/tls-certificates-interface/blob/d458e81e2f9941f1cb3601b50b5823436417a129/lib/charms/tls_certificates_interface/v4/tls_certificates.py#L675-L703)). I don't know how feasible this is, but by playing with it for a short bit I figure we'd probably need to create our own class that wraps the underlying objects, and it's a _much_ bigger change than fixing this for the tests. I know none of this is new to v4, but I'm curious if now is the time to tackle it.

My concern boils down to this: there are only two times we actually care about the string versions.

1. when we want to put it in the databag
2. when we want to write it to disk.

and never do we actually care about the bytes versions (it's only used as an intermediary as we're serializing and deserializing as far as I can tell). So those two things (string version, and bytes version) should probably never show up in any public functions with the exception of case 2.

Case 1 could be entirely encapsulated by the library, so a Charm author never needs to worry about it. Every time it is loaded from the databag, it could be converted directly into the object version, and vice versa when we write to the databag.

And case 2 should be a very simple operation on whatever object(s) we give them (possibly as simple as  a `str(cert)` operation).  

Checking if the CSR matches the private key, then, should be as simple as `csr.public_key() == private_key.public_key()` for the Charm author because they will always have object versions of these things -- although maybe it will be our objects, and so the method signatures might be slightly different.

I hope that made sense, but if not, lets touch base. Maybe I'm way off on this.

Reply:
The user won't have to manipulate the bytes version of the certificate anymore at all, any conversion is done in the tls library. You can see this in the requirer charm in the integration and unit tests. `generate_csr`, `generate_private_key` are all called inside the lib now.

Reply:
I did a couple of changes following this discussion:
- Return certificates as objects instead of strings in getters (ex. `get_assigned_certificate`, `get_csrs_from_requirer_relation_data`)
- Make databag classes private
- Make `generate` methods private

Let me know what you think.